File: 10. Miscellaneous Teasers/hanoi.c
#include <stdio.h>

#define Source 1
#define Middle 2
#define Target 3

int n = 10;
int count = 1;

void solve(int n, int src, int mdl, int tgt){
    if (n == 1){
        printf("Step %d : Move a ring from tower %d to %d\n", count++, src, tgt);
    } else {
        solve(n-1, src, tgt, mdl);
        solve(1, src, mdl, tgt);
        solve(n-1, mdl, src, tgt);
    }
}

void main(){
    solve(n, Source, Middle, Target);
}

File: Algorithms/Series Approximation & Recursion/hanoi.go
package main

import (
	"fmt"
)

var (
	n       = 8
	counter = 0
)

func main() {
	solve(n, 1, 2, 3)
}

func solve(n, source, tmp, target int) {
	if n == 1 {
		counter++
		move(source, target)
	} else if n > 1 {
		solve(n-1, source, target, tmp)
		solve(1, source, tmp, target)
		solve(n-1, tmp, source, target)
	}
}
func move(source, target int) {
	fmt.Printf("Step %d : Move a ring from tower %d to tower %d\n", counter, source, target)
}
